src/main/java/com/iexec/common/docker/client/DockerClientInstance.java

```diff
@@ -411,27 +413,41 @@ public DockerRunResponse run(DockerRunRequest dockerRunRequest) {
         if (dockerRunRequest.getMaxExecutionTime() <= 0) {
             // container will run until self-exited or explicitly-stopped
             log.info("Docker container will run in detached mode [name:{}]", containerName);
-            dockerRunResponse.setSuccessful(true);
+            dockerRunResponse.setFinalStatus(DockerRunFinalStatus.SUCCESS);
             return dockerRunResponse;
         }
         Instant timeoutDate = Instant.now()
                 .plusMillis(dockerRunRequest.getMaxExecutionTime());
-        int exitCode = waitContainerUntilExitOrTimeout(containerName, timeoutDate);
-        dockerRunResponse.setContainerExitCode(exitCode);
-        boolean isTimeout = exitCode == -1;
-        boolean isSuccessful = !isTimeout && exitCode == 0L;
-        if (isTimeout && !stopContainer(containerName)) {
-            log.error("Failed to force-stop container after timeout [name:{}]", containerName);
-            return dockerRunResponse;
+        boolean isSuccessful;
+        try {
+            int exitCode = waitContainerUntilExitOrTimeout(containerName, timeoutDate);
+            dockerRunResponse.setContainerExitCode(exitCode);
+
+            isSuccessful = exitCode == 0L;
+            log.info("Finished running docker container [name:{}, isSuccessful:{}]",
+                    containerName, isSuccessful);
+            dockerRunResponse.setFinalStatus(
+                    isSuccessful
+                            ? DockerRunFinalStatus.SUCCESS
+                            : DockerRunFinalStatus.FAILED);
+        } catch (TimeoutException e) {
+            log.error(e.getMessage());
+            dockerRunResponse.setFinalStatus(DockerRunFinalStatus.TIMEOUT);
+            if (!stopContainer(containerName)) {
```

Comment:
Something doesn't seem alright to me in this method.
I understand we retrieve logs then exit if `stopContainer` does not behave properly.
But with the current implementation, we return before calling the `removeContainer` method which happens during a normal execution.
Shouldn't we log the error and try to kill the container, then try to follow through the end of the code ?

Reply:
Don't we need a `stopContainer` method where we use this ?
Not-null timeout to have a grace period before a forceful kill
https://github.com/docker-java/docker-java/blob/master/docker-java-api/src/main/java/com/github/dockerjava/api/command/StopContainerCmd.java#L15